src/parsers.rs

```diff
@@ -1,22 +1,784 @@
+                let record = self.parse_date_time_utf16(source)?;
+                self.validate_and_build_date_time_utf16(record)
+            }
+        }
+    }
+
+    /// Parses the source into a `ZonedDateTime` compatible record.
+    pub fn parse_zoned_date_time(&self) -> TemporalResult<ParsedZonedDateTime<'a>> {
+        match &self.encoding {
+            ParserEncoding::Utf8(source) => {
+                let source_str = core::str::from_utf8(source)
+                    .map_err(|_| TemporalError::syntax().with_message("Invalid UTF-8 in source"))?;
+                let record = parse_zoned_date_time(source_str)?;
+                self.validate_and_build_zoned_date_time(record)
+            }
+            ParserEncoding::Utf16(source) => {
+                let record = self.parse_zoned_date_time_utf16(source)?;
+                self.validate_and_build_zoned_date_time_utf16(record)
+            }
+        }
+    }
+
+    /// Parses the source into an `Instant` compatible record.
+    pub fn parse_instant(&self) -> TemporalResult<ParsedInstant> {
+        let record = match &self.encoding {
+            ParserEncoding::Utf8(source) => parse_instant(source)?,
+            ParserEncoding::Utf16(source) => self.parse_instant_utf16(source)?,
+        };
+        self.validate_and_build_instant(record)
+    }
+
+    /// Parses the source into a `PlainTime` compatible record.
+    pub fn parse_time(&self) -> TemporalResult<ParsedTime> {
+        let record = match &self.encoding {
+            ParserEncoding::Utf8(source) => parse_time(source)?,
+            ParserEncoding::Utf16(source) => self.parse_time_utf16(source)?,
+        };
+        self.validate_and_build_time(record)
+    }
+
+    /// Parses the source into a `PlainYearMonth` compatible record.
+    pub fn parse_year_month(&self) -> TemporalResult<ParsedYearMonth<'a>> {
+        match &self.encoding {
+            ParserEncoding::Utf8(source) => {
+                let record = parse_year_month(source)?;
+                self.validate_and_build_year_month(record)
+            }
+            ParserEncoding::Utf16(source) => {
+                let record = self.parse_year_month_utf16(source)?;
+                self.validate_and_build_year_month_utf16(record)
+            }
+        }
+    }
+
+    /// Parses the source into a `PlainMonthDay` compatible record.
+    pub fn parse_month_day(&self) -> TemporalResult<ParsedMonthDay<'a>> {
+        match &self.encoding {
+            ParserEncoding::Utf8(source) => {
+                let record = parse_month_day(source)?;
+                self.validate_and_build_month_day(record)
+            }
+            ParserEncoding::Utf16(source) => {
+                let record = self.parse_month_day_utf16(source)?;
+                self.validate_and_build_month_day_utf16(record)
+            }
+        }
+    }
+
+    // Private UTF-16 parsing methods
+
+    fn parse_date_time_utf16(
+        &self,
+        source: &'a [u16],
```

Comment:
question: what's the point of taking both a `source` and `self` here? Both wrap the same string?